src/goal.rs

```diff
@@ -691,22 +644,45 @@ impl<'a> Goal<'a> {
         if (CONFIG.irreducible_only && self.is_reducible(rhs_expr)) || has_guard_wildcards(&rhs) {
           continue;
         }
-        let condition = SmallerVar {
-          scrutinees: self.scrutinees.iter().cloned().collect(),
-          // TODO: Can we take a reference instead of cloning?
-          ty_splits: self.ty_splits.clone(),
-          premises: self.premises.clone(),
+
+        let lhs_vars = var_set(&lhs);
+        let rhs_vars = var_set(&rhs);
+        let lemma_vars = lhs_vars.union(&rhs_vars).cloned().collect();
+
+        // If any of my premises contain variables that are not present in lhs or rhs,
+        // skip because we don't know how to check such a premise
+        if !premises.iter().all(|eq| {
+          let premise_lhs_vars = var_set(&to_pattern(&eq.lhs.expr, is_var));
+          let premise_rhs_vars = var_set(&to_pattern(&eq.rhs.expr, is_var));
+          let premise_vars: HashSet<Var> =
+            premise_lhs_vars.union(&premise_rhs_vars).cloned().collect();
+          premise_vars.is_subset(&lemma_vars)
+        }) {
+          continue;
+        }
+
+        // Pick out those variables that occur in the lemma
+        let lemma_var_classes: IdSubst = self
+          .var_classes
+          .iter()
+          .filter(|(x, _)| lemma_vars.contains(&to_wildcard(x)))
+          .map(|(x, id)| (x.clone(), *id))
+          .collect();
+
+        let condition = Soundness {
+          free_vars: lemma_var_classes,
+          premises: premises.clone(),
         };
         let mut added_lemma = false;
-        if rhs.vars().iter().all(|x| lhs.vars().contains(x)) {
+        if rhs_vars.is_subset(&lhs_vars) {
           // if rhs has no extra wildcards, create a lemma lhs => rhs
           Goal::add_lemma(lhs.clone(), rhs.clone(), condition.clone(), &mut rewrites);
           added_lemma = true;
           if CONFIG.single_rhs {
             continue;
           };
         }
-        if (is_cyclic || !added_lemma) && lhs.vars().iter().all(|x| rhs.vars().contains(x)) {
+        if (is_cyclic || !added_lemma) && lhs_vars.is_subset(&rhs_vars) {
           // if lhs has no extra wildcards, create a lemma rhs => lhs;
           // in non-cyclic mode, a single direction of IH is always sufficient
           // (because grounding adds all instantiations we could possibly care about).
```

Comment:
Okay, so what I meant here is this:

- If both LHS => RHS and RHS => LHS are valid rewrites, we never need to have them both in the presence of grounding. This is because grounding adds both LHS and RHS terms to the e-graph (for all sound instantiations), so neither rule can create new terms, and all it will do will just union the two e-classes, and hence there's no point having both rules.
- However, what to do if it so happens that LHS => RHS is not a valid rewrite (i.e. RHS has more variables)? Then I would argue that we should reject such a prop outright (as well as a prop where the condition has extra arguments).

Reply:
Hmmm, I think we should maybe revisit this discussion in the future in case this precondition no longer holds. Especially if we end up with laxer soundness checks via blocking var analysis.

Reply:
Well, if we have a laxer soundness check, which will make grounding incomplete, we'll have a bigger problem than just this. In that case, allowing the IH to apply both ways would be a mere band-aid on the problem....